src/explore-education-statistics-common/src/modules/table-tool/utils/__tests__/applyTableHeadersOrder.test.ts

```diff
@@ -0,0 +1,1542 @@
+      });
+      expect(result).toEqual(expected);
+    });
+  });
+
+  describe('mixing up rows and columns', () => {
+    const testReorderedHeadersForRowsAndColumns: TableHeadersConfig = {
+      columnGroups: [
+        testCategoryFilters3,
+        testCategoryFilters1,
+        testCategoryFilters2,
+      ],
+      columns: testTimePeriodFilters,
+      rowGroups: [testLocationFilters],
+      rows: testIndicators,
+    };
+    test('retains the order when a rowGroup has been moved to columnGroups and no options have changed', () => {
+      const testUpdatedTableHeaders: TableHeadersConfig = {
+        columnGroups: [testCategoryFilters1, testCategoryFilters2],
+        columns: testTimePeriodFilters,
+        rowGroups: [testCategoryFilters3, testLocationFilters],
+        rows: testIndicators,
+      };
+      const result = applyTableHeadersOrder({
+        reorderedTableHeaders: testReorderedHeadersForRowsAndColumns,
+        updatedTableHeaders: testUpdatedTableHeaders,
+      });
+      expect(result).toEqual(testReorderedHeadersForRowsAndColumns);
+    });
+
+    test('retains the order when a rowGroup has been moved to columnGroups and options have been added and removed', () => {
+      const testUpdatedTableHeaders: TableHeadersConfig = {
+        columnGroups: [testCategoryFilters1],
+        columns: testTimePeriodFilters,
+        rowGroups: [
+          testCategoryFilters3,
+          testLocationFilters,
+          testCategoryFilters4,
+        ],
+        rows: testIndicators,
+      };
+      const expected: TableHeadersConfig = {
+        columnGroups: [testCategoryFilters3, testCategoryFilters1],
+        columns: testTimePeriodFilters,
+        rowGroups: [testLocationFilters, testIndicators],
+        rows: testCategoryFilters4,
+      };
+      const result = applyTableHeadersOrder({
+        reorderedTableHeaders: testReorderedHeadersForRowsAndColumns,
+        updatedTableHeaders: testUpdatedTableHeaders,
+      });
+      expect(result).toEqual(expected);
+    });
+
+    test('retains the order when rows has been moved to columnGroups and options have been added and removed', () => {
+      const testReorderedHeaders: TableHeadersConfig = {
+        columnGroups: [
+          testIndicators,
+          testCategoryFilters1,
+          testCategoryFilters2,
+        ],
+        columns: testTimePeriodFilters,
+        rowGroups: [testLocationFilters],
+        rows: testCategoryFilters3,
+      };
+      const testUpdatedTableHeaders: TableHeadersConfig = {
+        columnGroups: [
+          testCategoryFilters1,
+          testCategoryFilters2,
+          testTimePeriodFilters,
+        ],
+        columns: testCategoryFilters4,
+        rowGroups: [testLocationFilters],
+        rows: testIndicators,
+      };
+      const expected: TableHeadersConfig = {
+        columnGroups: [
+          testIndicators,
+          testCategoryFilters1,
+          testCategoryFilters2,
+          testTimePeriodFilters,
```

Comment:
I think the backend does this optimisation already when it generates the table headers so I don't think it's necessary to add complexity here to move things around that haven't been reordered by users.

Reply:
Yeah don't know what I was thinking there! Obviously it's `getDefaultTableHeadersConfig`.

I've added something fairly simple to populate the rows with new column groups if there would otherwise only have been one row group. We might want to expand on this in future to add more scenarios for when to move them.